testnet/stacks-node/src/burnchains/db_indexer.rs

```diff
@@ -0,0 +1,589 @@
+    connection: &DBConn,
+    last_canonical_tip: BurnchainHeaderHash,
+) -> Result<u64, BurnchainError> {
+    let mut cursor = last_canonical_tip;
+    loop {
+        let cursor_header = match query_row::<BurnHeaderDBRow, _>(
+            connection,
+            "SELECT * FROM headers WHERE header_hash = ?1",
+            &[&cursor],
+        )? {
+            Some(header) => header,
+            None => {
+                // TODO: Should this be an error?
+                panic!("Do we hit here?");
+            }
+        };
+
+        if cursor_header.is_canonical != 0 {
+            return Ok(cursor_header.height);
+        }
+
+        cursor = cursor_header.parent_header_hash;
+    }
+}
+
+impl BurnchainChannel for DBBurnBlockInputChannel {
+    fn push_block(&self, new_block: NewBlock) -> Result<(), BurnchainError> {
+        // Re-open the connection.
+        let open_flags = OpenFlags::SQLITE_OPEN_READ_WRITE;
+        let connection = sqlite_open(&self.output_db_path, open_flags, true)?;
+
+        // Decide if this new node is part of the canonical chain.
+        let current_canonical_tip_opt = get_canonical_chain_tip(&connection);
+
+        let header = BurnHeaderDBRow::from(&new_block);
+        let (is_canonical, needs_reorg) = match &current_canonical_tip_opt {
+            // No canonical tip so no re-org.
+            None => (true, false),
+
+            Some(current_canonical_tip) => {
+                // `new_blocks` parent is the old tip, so no reorg.
+                if header.parent_header_hash() == current_canonical_tip.header_hash() {
+                    (true, false)
+                } else {
+                    // `new_block` isn't the child of the current tip. We ASSUME we have seen all blocks before now.
+                    // So, this must be a different chain. Check to see if this is a longer tip.
+                    let compare_result = compare_headers(current_canonical_tip, &header);
+                    if compare_result > 0 {
+                        // The new block is greater than the previous tip. It is canonical, and we need a reorg.
+                        (true, true)
+                    } else {
+                        (false, false)
+                    }
+                }
+            }
+        };
+
+        // Insert this header.
+        let params: &[&dyn ToSql] = &[
+            &(header.height() as u32),
+            &BurnchainHeaderHash(header.header_hash()),
+            &BurnchainHeaderHash(header.parent_header_hash()),
+            &(header.time_stamp() as u32),
+            &(is_canonical as u32),
+        ];
+        match connection.execute(
+            "INSERT INTO headers (height, header_hash, parent_header_hash, time_stamp, is_canonical) VALUES (?, ?, ?, ?, ?)",
+            params,
+        ) {
+            Ok(_) => {            }
+            Err(e) => {
+                return Err(BurnchainError::DBError(db_error::SqliteError(e)));
+            }
+        };
```

Comment:
Because `BurnchainError` implements `From<RusqliteError>`, this can be written as:

```suggestion
        connection.execute(
            "INSERT INTO headers (height, header_hash, parent_header_hash, time_stamp, is_canonical) VALUES (?, ?, ?, ?, ?)",
            params,
        )?;
```